crates/eframe/src/web/web_painter_wgpu.rs

```diff
@@ -83,46 +83,95 @@ impl WebPainterWgpu {
                 log::info!(
-                    "WebGPU is only available in secure contexts, i.e. on HTTPS and on localhost"
+                    "WebGPU is only available in secure contexts, i.e. on HTTPS and on localhost."
                 );
+
+                // Don't try WebGPU since we established now that it will fail.
+                backends.remove(wgpu::Backends::BROWSER_WEBGPU);
+
+                if backends.is_empty() {
+                    return Err("No available supported graphics backends.".to_owned());
+                }
             }
         }
 
         let instance = wgpu::Instance::new(wgpu::InstanceDescriptor {
-            backends: options.wgpu_options.supported_backends,
+            backends,
             ..Default::default()
         });
-
-        let canvas = super::canvas_element_or_die(canvas_id);
-
         let surface = instance
             .create_surface(wgpu::SurfaceTarget::Canvas(canvas.clone()))
             .map_err(|err| format!("failed to create wgpu surface: {err}"))?;
+        let render_state_result =
+            RenderState::create(wgpu_options, &instance, &surface, depth_format, 1).await;
+
+        if let Err(egui_wgpu::WgpuError::NoSuitableAdapterFound) = &render_state_result {
+            if backends == wgpu::Backends::BROWSER_WEBGPU | wgpu::Backends::GL {
+                // It can happen that a browser advertises WebGPU support, but then fails to create a
+                // suitable adapter. As of writing this happens for example on Linux with Chrome 121.
+                //
+                // Since WebGPU is handled in a special way in wgpu, we have to recreate the instance
+                // if we instead want to try with WebGL.
+                //
+                // Note that we also might end up here if wgpu didn't detect WebGPU support,
+                // moved on to create a wgpu-core instance and then failed to create a WebGL adapter.
+                // To detect this we'd need to check what kind of instance wgpu created,
+                // but this is currently not easily possible. See https://github.com/gfx-rs/wgpu/issues/5142
+                // In that case we'd be trying the same thing again which isn't ideal, but we're in a
+                // pretty bad situation anyway.
+                let instance = wgpu::Instance::new(wgpu::InstanceDescriptor {
+                    backends: wgpu::Backends::GL,
+                    ..Default::default()
+                });
+                let surface = instance
+                    .create_surface(wgpu::SurfaceTarget::Canvas(canvas.clone()))
+                    .map_err(|err| format!("failed to create wgpu surface: {err}"))?;
+                let render_state =
+                    RenderState::create(wgpu_options, &instance, &surface, depth_format, 1)
+                        .await
+                        .map_err(|err| err.to_string())?;
+
+                return Ok((render_state, surface));
```

Comment:
isn't it easier to just recurse here?

```suggestion
                return Sellf::init_wgpu_and_create_render_state(
                    wgpu::Backends::GL,
                    wgpu_options,
                    depth_format,
                    canvas,
                );
                    
```

Reply:
I wanted to do that originally but async functions are apparently not allowed to recurse. After a bit back and forth I figured it's easiest to just copy & paste those few lines